src/effects/hearts/index.ts

```diff
@@ -0,0 +1,190 @@
+    maxRot: 5,
+};
+
+const KEY_FRAME_INTERVAL = 15; // 15ms, roughly
+
+export default class Hearts implements ICanvasEffect {
+    private readonly options: HeartOptions;
+
+    constructor(options: { [key: string]: any }) {
+        this.options = { ...DefaultOptions, ...options };
+    }
+
+    private context: CanvasRenderingContext2D | null = null;
+    private particles: Array<Heart> = [];
+    private lastAnimationTime: number;
+
+    private colours = [
+        'rgba(194,210,224,1)',
+        'rgba(235,214,219,1)',
+        'rgba(255,211,45,1)',
+        'rgba(255,190,174,1)',
+        'rgba(255,173,226,1)',
+        'rgba(242,114,171,1)',
+        'rgba(228,55,116,1)',
+        'rgba(255,86,130,1)',
+        'rgba(244,36,57,1)',
+        'rgba(247,126,157,1)',
+        'rgba(243,142,140,1)',
+        'rgba(252,116,183,1)'];
+
+    public isRunning: boolean;
+
+    public start = async (canvas: HTMLCanvasElement, timeout = 3000) => {
+        if (!canvas) {
+            return;
+        }
+        this.context = canvas.getContext('2d');
+        this.particles = [];
+        const count = this.options.maxCount;
+        while (this.particles.length < count) {
+            this.particles.push(this.resetParticle({} as Heart, canvas.width, canvas.height));
+        }
+        this.isRunning = true;
+        requestAnimationFrame(this.renderLoop);
+        if (timeout) {
+            window.setTimeout(this.stop, timeout);
+        }
+    };
+
+    public stop = async () => {
+        this.isRunning = false;
+    };
+
+    private resetParticle = (particle: Heart, width: number, height: number): Heart => {
+        particle.color = this.colours[(Math.random() * this.colours.length) | 0];
+        particle.x = Math.random() * width;
+        particle.y = Math.random() * height + height;
+        particle.xCol = particle.x;
+        particle.scale = (Math.random() * 0.07) + 0.04;
+        particle.maximumDrift = (Math.random() * this.options.maxDrift) + 3.5;
+        particle.maximumRot = (Math.random() * this.options.maxRot) + 3.5;
+        particle.gravity = this.options.gravity + (Math.random() * 4.8);
+        return particle;
+    };
+
+    private renderLoop = (): void => {
+        if (!this.context || !this.context.canvas) {
+            return;
+        }
+        if (this.particles.length === 0) {
+            this.context.clearRect(0, 0, this.context.canvas.width, this.context.canvas.height);
+        } else {
+            const timeDelta = Date.now() - this.lastAnimationTime;
+            if (timeDelta >= KEY_FRAME_INTERVAL || !this.lastAnimationTime) {
+                // Clear the screen first
+                this.context.clearRect(0, 0, this.context.canvas.width, this.context.canvas.height);
+
+                this.lastAnimationTime = Date.now();
+                this.animateAndRenderHearts();
+            }
+            requestAnimationFrame(this.renderLoop);
+        }
+    };
+
+    private animateAndRenderHearts() {
```

Comment:
If we're adding more chat effects by copying the existing ones, I'd really rather factor out the common code than copy & paste it between files (this is the same comment from the snowfall code).

Reply:
Are you sure it would be a good idea to refactor all the existing effects as part of this PR? Other effects, most recently the rainfall effect in Nov 2021 ( https://github.com/matrix-org/matrix-react-sdk/pull/7086 ) have been merged without any issues raised about copied code.

Reply:
It's a long-term concern from before the rainfall effect as well. In general we aim for each PR to be better than the last, improving the situation as we go.

Given this effect is one of the last we'll be merging though, I'm not overly concerned with code duplication personally. The area could use work, but that can be a different PR. If it was included in this PR, all the better.